### PR TITLE
Isolate petsc4y in Python interface

### PR DESCRIPTION
**Pull request by garth-wells**
Also:
- Fixed a number of `dolfinx::la::Vector` functions for complex types.
- Python documentation improvements.

### CODE REVIEW

File: python/dolfinx/io/utils.py [hunk 1]
```diff
@@ -39,14 +39,13 @@ def _extract_cpp_functions(functions: typing.Union[typing.List[Function], Functi
     __all__ = __all__ + ["FidesWriter", "VTXWriter"]
 
     class VTXWriter:
-        """Interface to VTK files for ADIOS2
+        """Writer to to VTK files, using ADIOS2 to create the files.
```
Comment (chrisrichardson) [inline python/dolfinx/io/utils.py:42]:
```suggestion
        """Writer for VTK files, using ADIOS2 to create the files.
```

File: python/dolfinx/io/utils.py [hunk 2]
```diff
@@ -100,15 +99,14 @@ def close(self):
             self._cpp_object.close()
 
     class FidesWriter:
-        """Interface to Fides file format.
+        """Writer to to Fides files, using ADIOS2 to create the files.
```
Comment (chrisrichardson) [inline python/dolfinx/io/utils.py:102]:
```suggestion
        """Writer for Fides files, using ADIOS2 to create the files.
```

File: python/dolfinx/io/utils.py [hunk 3]
```diff
@@ -100,15 +99,14 @@ def close(self):
...
 
-        Fides supports first order Lagrange finite elements for the
-        geometry description and first order Lagrange finite elements
-        for functions. All functions has to be of the same element
-        family and same order.
+        Fides (https://fides.readthedocs.io/) supports first order
+        Lagrange finite elements for the geometry description and first
+        order Lagrange finite elements for functions. All functions has
```
Comment (chrisrichardson) [inline python/dolfinx/io/utils.py:106]:
```suggestion
        order Lagrange finite elements for functions. All functions have
```

File: python/dolfinx/plot.py [hunk 4]
```diff
@@ -83,9 +91,21 @@ def create_vtk_mesh(msh: mesh.Mesh, dim: typing.Optional[int] = None, entities=N
 @create_vtk_mesh.register(fem.FunctionSpace)
 def _(V: fem.FunctionSpace, entities=None):
     """Creates a VTK mesh topology (topology array and array of cell
-    types) that is based on the degree-of-freedom coordinates. Note that
-    this function supports Lagrange elements (continuous and
-    discontinuous) only.
+    types) that is based on the degree-of-freedom coordinates.
+
+    This function support visualisation when the degree of the finite
```
Comment (chrisrichardson) [inline python/dolfinx/plot.py:96]:
```suggestion
    This function supports visualisation when the degree of the finite
```

**Review by chrisrichardson — APPROVED** [review]

File: python/dolfinx/la.py [hunk 5]
```diff
@@ -232,14 +250,34 @@ def create_petsc_vector_wrap(x: Vector):
...
-    return _cpp.la.petsc.create_vector_wrap(x._cpp_object)
+    from petsc4py import PETSc
+    map = x.index_map
+    ghosts = map.ghosts.astype(PETSc.IntType)
+    bs = x.block_size
+    size = (map.size_local * bs, map.size_global * bs)
+    return PETSc.Vec().createGhostWithArray(ghosts, x.array, size=size, bsize=bs, comm=map.comm)
+
+
+def create_petsc_vector(map, bs: int):
```
Comment (jorgensd) [inline python/dolfinx/la.py:268]: This should really have a type hint (`dolfinx.common.IndexMap`)?

**Pull request merged**